node/src/daser.rs

```diff
@@ -110,74 +125,290 @@ where
             .await?;
 
-        Ok(())
-    }
+        let p2p = self.p2p.clone();
 
-    async fn sample_block(&mut self, header: &ExtendedHeader) -> Result<(Vec<Cid>, bool)> {
-        let now = Instant::now();
-        let indexes = random_indexes(header.dah.square_width(), self.max_samples_needed);
-        let mut futs = FuturesUnordered::new();
+        // Schedule retrival of the CIDs. This will be run later on in the `select!` loop.
+        self.sampling_fut.set(async move {
+            let now = Instant::now();
+            let mut futs = FuturesUnordered::new();
 
-        for (row_index, column_index) in indexes {
-            let fut = self
-                .p2p
-                .get_sample(row_index, column_index, header.height().value());
-            futs.push(fut);
-        }
+            for cid in cids {
+                let fut = p2p.get_shwap_cid(cid, Some(GET_SAMPLE_TIMEOUT));
+                futs.push(fut);
+            }
+
+            let mut accepted = true;
+
+            while let Some(res) = futs.next().await {
+                match res {
+                    Ok(_) => {}
+                    // Validation is done at Bitswap level, through `ShwapMultihasher`.
+                    // If the sample is not valid, it will never be delivered to us
+                    // as the data of the CID. Because of that, the only signal
+                    // that data sampling verification failed is query timing out.
+                    Err(P2pError::BitswapQueryTimeout) => accepted = false,
+                    Err(e) => return Err(e.into()),
+                }
+            }
+
+            debug!(
+                "Data sampling of {} is {}. Took {:?}",
+                args.height,
+                if accepted { "accepted" } else { "rejected" },
+                now.elapsed()
+            );
+
+            Ok((args.height, accepted))
+        });
+
+        Ok(())
+    }
 
-        let mut cids: Vec<Cid> = Vec::new();
-        let mut accepted = true;
-
-        while let Some(res) = futs.next().await {
-            match res {
-                Ok(sample) => cids.push(convert_cid(&sample.id.into())?),
-                // Validation is done at Bitswap level, through `ShwapMultihasher`.
-                // If the sample is not valid, it will never be delivered to us
-                // as the data of the CID. Because of that, the only signal
-                // that data sampling verification failed is query timing out.
-                Err(P2pError::BitswapQueryTimeout) => accepted = false,
-                Err(e) => return Err(e.into()),
+    /// Add to the queue the blocks that need to be sampled.
+    ///
+    /// NOTE: We resampled rejected blocks, because rejection can happen
+    /// because of edge-cases, such us network issues. This is a limitation
+    /// because of how Shwap works: The only way to know if sampling has failed
+    /// is via a timeout.
```

Comment:
```suggestion
    /// NOTE: We resample rejected blocks, because rejection can happen
    /// in some unrelated edge-cases, such us network issues. This is a Shwap 
    /// limitation that's coming from bitswap: only way for us to know if sampling
    /// failed is via timeout.
```